Build fail due to undeclared isnan/rnan

I noticed build failure on Ubuntu 16.04 with GCC 5.3.1 and Clang 3.8.0 too.

Clang build log:
--------------------
In file included from /home/ossy/wasm/binaryen/src/support/command-line.cpp:17:
In file included from /home/ossy/wasm/binaryen/src/support/command-line.h:30:
/home/ossy/wasm/binaryen/src/wasm.h:236:9: error: use of undeclared identifier 'isnan'
    if (isnan(f)) {
        ^
/home/ossy/wasm/binaryen/src/wasm.h:252:9: error: use of undeclared identifier 'isnan'
    if (isnan(d)) {
        ^
/home/ossy/wasm/binaryen/src/wasm.h:631:21: error: use of undeclared identifier 'isnan'
        bool lnan = isnan(l), rnan = isnan(r);
                    ^
/home/ossy/wasm/binaryen/src/wasm.h:631:38: error: use of undeclared identifier 'isnan'
        bool lnan = isnan(l), rnan = isnan(r);
                                     ^
/home/ossy/wasm/binaryen/src/wasm.h:632:14: error: use of undeclared identifier 'isnan'
        if (!isnan(result) && !lnan && !rnan) return Literal(result);
             ^
/home/ossy/wasm/binaryen/src/wasm.h:640:21: error: use of undeclared identifier 'isnan'
        bool lnan = isnan(l), rnan = isnan(r);
                    ^
/home/ossy/wasm/binaryen/src/wasm.h:640:38: error: use of undeclared identifier 'isnan'
        bool lnan = isnan(l), rnan = isnan(r);
                                     ^
/home/ossy/wasm/binaryen/src/wasm.h:641:14: error: use of undeclared identifier 'isnan'
        if (!isnan(result) && !lnan && !rnan) return Literal(result);
             ^
/home/ossy/wasm/binaryen/src/wasm.h:654:21: error: use of undeclared identifier 'isnan'
        bool lnan = isnan(l), rnan = isnan(r);
                    ^
/home/ossy/wasm/binaryen/src/wasm.h:654:38: error: use of undeclared identifier 'isnan'
        bool lnan = isnan(l), rnan = isnan(r);
                                     ^
/home/ossy/wasm/binaryen/src/wasm.h:655:14: error: use of undeclared identifier 'isnan'
        if (!isnan(result) && !lnan && !rnan) return Literal(result);
             ^
/home/ossy/wasm/binaryen/src/wasm.h:663:21: error: use of undeclared identifier 'isnan'
        bool lnan = isnan(l), rnan = isnan(r);
                    ^
/home/ossy/wasm/binaryen/src/wasm.h:663:38: error: use of undeclared identifier 'isnan'
        bool lnan = isnan(l), rnan = isnan(r);
                                     ^
/home/ossy/wasm/binaryen/src/wasm.h:664:14: error: use of undeclared identifier 'isnan'
        if (!isnan(result) && !lnan && !rnan) return Literal(result);
             ^
14 errors generated.


GCC build log:
--------------------
In file included from /home/ossy/wasm/binaryen/src/support/command-line.h:30:0,
                 from /home/ossy/wasm/binaryen/src/support/command-line.cpp:17:
/home/ossy/wasm/binaryen/src/wasm.h: In static member function ‘static void wasm::Literal::printFloat(std::ostream&, float)’:
/home/ossy/wasm/binaryen/src/wasm.h:236:16: error: ‘isnan’ was not declared in this scope
     if (isnan(f)) {
                ^
/home/ossy/wasm/binaryen/src/wasm.h:236:16: note: suggested alternative:
In file included from /home/ossy/wasm/binaryen/src/wasm.h:47:0,
                 from /home/ossy/wasm/binaryen/src/support/command-line.h:30,
                 from /home/ossy/wasm/binaryen/src/support/command-line.cpp:17:
/usr/include/c++/5/cmath:641:5: note:   ‘std::isnan’
     isnan(_Tp __x)
     ^
In file included from /home/ossy/wasm/binaryen/src/support/command-line.h:30:0,
                 from /home/ossy/wasm/binaryen/src/support/command-line.cpp:17:
/home/ossy/wasm/binaryen/src/wasm.h: In static member function ‘static void wasm::Literal::printDouble(std::ostream&, double)’:
/home/ossy/wasm/binaryen/src/wasm.h:252:16: error: ‘isnan’ was not declared in this scope
     if (isnan(d)) {
                ^
/home/ossy/wasm/binaryen/src/wasm.h:252:16: note: suggested alternative:
In file included from /home/ossy/wasm/binaryen/src/wasm.h:47:0,
                 from /home/ossy/wasm/binaryen/src/support/command-line.h:30,
                 from /home/ossy/wasm/binaryen/src/support/command-line.cpp:17:
/usr/include/c++/5/cmath:641:5: note:   ‘std::isnan’
     isnan(_Tp __x)
     ^
In file included from /home/ossy/wasm/binaryen/src/support/command-line.h:30:0,
                 from /home/ossy/wasm/binaryen/src/support/command-line.cpp:17:
/home/ossy/wasm/binaryen/src/wasm.h: In member function ‘wasm::Literal wasm::Literal::min(const wasm::Literal&) const’:
/home/ossy/wasm/binaryen/src/wasm.h:631:28: error: ‘isnan’ was not declared in this scope
         bool lnan = isnan(l), rnan = isnan(r);
                            ^
/home/ossy/wasm/binaryen/src/wasm.h:631:28: note: suggested alternative:
In file included from /home/ossy/wasm/binaryen/src/wasm.h:47:0,
                 from /home/ossy/wasm/binaryen/src/support/command-line.h:30,
                 from /home/ossy/wasm/binaryen/src/support/command-line.cpp:17:
/usr/include/c++/5/cmath:641:5: note:   ‘std::isnan’
     isnan(_Tp __x)
     ^
In file included from /home/ossy/wasm/binaryen/src/support/command-line.h:30:0,
                 from /home/ossy/wasm/binaryen/src/support/command-line.cpp:17:
/home/ossy/wasm/binaryen/src/wasm.h:632:41: error: ‘rnan’ was not declared in this scope
         if (!isnan(result) && !lnan && !rnan) return Literal(result);
                                         ^
/home/ossy/wasm/binaryen/src/wasm.h:633:23: error: ‘rnan’ was not declared in this scope
         if (!lnan && !rnan) return Literal((int32_t)0x7fc00000).castToF32();
                       ^
/home/ossy/wasm/binaryen/src/wasm.h:640:28: error: ‘isnan’ was not declared in this scope
         bool lnan = isnan(l), rnan = isnan(r);
                            ^
/home/ossy/wasm/binaryen/src/wasm.h:640:28: note: suggested alternative:
In file included from /home/ossy/wasm/binaryen/src/wasm.h:47:0,
                 from /home/ossy/wasm/binaryen/src/support/command-line.h:30,
                 from /home/ossy/wasm/binaryen/src/support/command-line.cpp:17:
/usr/include/c++/5/cmath:641:5: note:   ‘std::isnan’
     isnan(_Tp __x)
     ^
In file included from /home/ossy/wasm/binaryen/src/support/command-line.h:30:0,
                 from /home/ossy/wasm/binaryen/src/support/command-line.cpp:17:
/home/ossy/wasm/binaryen/src/wasm.h:641:41: error: ‘rnan’ was not declared in this scope
         if (!isnan(result) && !lnan && !rnan) return Literal(result);
                                         ^
/home/ossy/wasm/binaryen/src/wasm.h:642:23: error: ‘rnan’ was not declared in this scope
         if (!lnan && !rnan) return Literal((int64_t)0x7ff8000000000000LL).castToF64();
                       ^
/home/ossy/wasm/binaryen/src/wasm.h: In member function ‘wasm::Literal wasm::Literal::max(const wasm::Literal&) const’:
/home/ossy/wasm/binaryen/src/wasm.h:654:28: error: ‘isnan’ was not declared in this scope
         bool lnan = isnan(l), rnan = isnan(r);
                            ^
/home/ossy/wasm/binaryen/src/wasm.h:654:28: note: suggested alternative:
In file included from /home/ossy/wasm/binaryen/src/wasm.h:47:0,
                 from /home/ossy/wasm/binaryen/src/support/command-line.h:30,
                 from /home/ossy/wasm/binaryen/src/support/command-line.cpp:17:
/usr/include/c++/5/cmath:641:5: note:   ‘std::isnan’
     isnan(_Tp __x)
     ^
In file included from /home/ossy/wasm/binaryen/src/support/command-line.h:30:0,
                 from /home/ossy/wasm/binaryen/src/support/command-line.cpp:17:
/home/ossy/wasm/binaryen/src/wasm.h:655:41: error: ‘rnan’ was not declared in this scope
         if (!isnan(result) && !lnan && !rnan) return Literal(result);
                                         ^
/home/ossy/wasm/binaryen/src/wasm.h:656:23: error: ‘rnan’ was not declared in this scope
         if (!lnan && !rnan) return Literal((int32_t)0x7fc00000).castToF32();
                       ^
/home/ossy/wasm/binaryen/src/wasm.h:663:28: error: ‘isnan’ was not declared in this scope
         bool lnan = isnan(l), rnan = isnan(r);
                            ^
/home/ossy/wasm/binaryen/src/wasm.h:663:28: note: suggested alternative:
In file included from /home/ossy/wasm/binaryen/src/wasm.h:47:0,
                 from /home/ossy/wasm/binaryen/src/support/command-line.h:30,
                 from /home/ossy/wasm/binaryen/src/support/command-line.cpp:17:
/usr/include/c++/5/cmath:641:5: note:   ‘std::isnan’
     isnan(_Tp __x)
     ^
In file included from /home/ossy/wasm/binaryen/src/support/command-line.h:30:0,
                 from /home/ossy/wasm/binaryen/src/support/command-line.cpp:17:
/home/ossy/wasm/binaryen/src/wasm.h:664:41: error: ‘rnan’ was not declared in this scope
         if (!isnan(result) && !lnan && !rnan) return Literal(result);
                                         ^
/home/ossy/wasm/binaryen/src/wasm.h:665:23: error: ‘rnan’ was not declared in this scope
         if (!lnan && !rnan) return Literal((int64_t)0x7ff8000000000000LL).castToF64();
                       ^

--- a/src/parsing.h
+++ b/src/parsing.h
@@ -79,7 +79,7 @@ Expression* parseConst(cashew::IString s, WasmType type, MixedArena& allocator) 
             pattern = 0x7fc00000U;
           }
           if (negative) pattern |= 0x80000000U;
-          if (!isnan(bit_cast<float>(pattern))) pattern |= 1U;
+          if (!std::isnan(bit_cast<float>(pattern))) pattern |= 1U;
           ret->value = Literal(pattern).castToF32();
           break;
         }
@@ -93,7 +93,7 @@ Expression* parseConst(cashew::IString s, WasmType type, MixedArena& allocator) 
             pattern = 0x7ff8000000000000UL;
           }
           if (negative) pattern |= 0x8000000000000000ULL;
-          if (!isnan(bit_cast<double>(pattern))) pattern |= 1ULL;
+          if (!std::isnan(bit_cast<double>(pattern))) pattern |= 1ULL;
           ret->value = Literal(pattern).castToF64();
           break;
         }

--- a/src/wasm-interpreter.h
+++ b/src/wasm-interpreter.h
@@ -636,7 +636,7 @@ private:
 
       Literal truncSFloat(Unary* curr, Literal value) {
         double val = value.getFloat();
-        if (isnan(val)) trap("truncSFloat of nan");
+        if (std::isnan(val)) trap("truncSFloat of nan");
         if (curr->type == i32) {
           if (val > (double)std::numeric_limits<int32_t>::max() || val < (double)std::numeric_limits<int32_t>::min()) trap("i32.truncSFloat overflow");
           return Literal(int32_t(val));
@@ -649,7 +649,7 @@ private:
 
       Literal truncUFloat(Unary* curr, Literal value) {
         double val = value.getFloat();
-        if (isnan(val)) trap("truncUFloat of nan");
+        if (std::isnan(val)) trap("truncUFloat of nan");
         if (curr->type == i32) {
           if (val > (double)std::numeric_limits<uint32_t>::max() || val <= (double)-1) trap("i32.truncUFloat overflow");
           return Literal(uint32_t(val));

--- a/src/wasm.h
+++ b/src/wasm.h
@@ -233,7 +233,7 @@ public:
   }
 
   static void printFloat(std::ostream &o, float f) {
-    if (isnan(f)) {
+    if (std::isnan(f)) {
       const char *sign = std::signbit(f) ? "-" : "";
       o << sign << "nan";
       if (uint32_t payload = ~0xff800000u & bit_cast<uint32_t>(f)) {
@@ -249,7 +249,7 @@ public:
       o << "-0";
       return;
     }
-    if (isnan(d)) {
+    if (std::isnan(d)) {
       const char *sign = std::signbit(d) ? "-" : "";
       o << sign << "nan";
       if (uint64_t payload = ~0xfff0000000000000ull & bit_cast<uint64_t>(d)) {
@@ -628,8 +628,8 @@ public:
         auto l = getf32(), r = other.getf32();
         if (l == r && l == 0) return Literal(std::signbit(l) ? l : r);
         auto result = std::min(l, r);
-        bool lnan = isnan(l), rnan = isnan(r);
-        if (!isnan(result) && !lnan && !rnan) return Literal(result);
+        bool lnan = std::isnan(l), rnan = std::isnan(r);
+        if (!std::isnan(result) && !lnan && !rnan) return Literal(result);
         if (!lnan && !rnan) return Literal((int32_t)0x7fc00000).castToF32();
         return Literal(lnan ? l : r).castToI32().or_(Literal(0xc00000)).castToF32();
       }
@@ -637,8 +637,8 @@ public:
         auto l = getf64(), r = other.getf64();
         if (l == r && l == 0) return Literal(std::signbit(l) ? l : r);
         auto result = std::min(l, r);
-        bool lnan = isnan(l), rnan = isnan(r);
-        if (!isnan(result) && !lnan && !rnan) return Literal(result);
+        bool lnan = std::isnan(l), rnan = std::isnan(r);
+        if (!std::isnan(result) && !lnan && !rnan) return Literal(result);
         if (!lnan && !rnan) return Literal((int64_t)0x7ff8000000000000LL).castToF64();
         return Literal(lnan ? l : r).castToI64().or_(Literal(int64_t(0x8000000000000LL))).castToF64();
       }
@@ -651,8 +651,8 @@ public:
         auto l = getf32(), r = other.getf32();
         if (l == r && l == 0) return Literal(std::signbit(l) ? r : l);
         auto result = std::max(l, r);
-        bool lnan = isnan(l), rnan = isnan(r);
-        if (!isnan(result) && !lnan && !rnan) return Literal(result);
+        bool lnan = std::isnan(l), rnan = std::isnan(r);
+        if (!std::isnan(result) && !lnan && !rnan) return Literal(result);
         if (!lnan && !rnan) return Literal((int32_t)0x7fc00000).castToF32();
         return Literal(lnan ? l : r).castToI32().or_(Literal(0xc00000)).castToF32();
       }
@@ -660,8 +660,8 @@ public:
         auto l = getf64(), r = other.getf64();
         if (l == r && l == 0) return Literal(std::signbit(l) ? r : l);
         auto result = std::max(l, r);
-        bool lnan = isnan(l), rnan = isnan(r);
-        if (!isnan(result) && !lnan && !rnan) return Literal(result);
+        bool lnan = std::isnan(l), rnan = std::isnan(r);
+        if (!std::isnan(result) && !lnan && !rnan) return Literal(result);
         if (!lnan && !rnan) return Literal((int64_t)0x7ff8000000000000LL).castToF64();
         return Literal(lnan ? l : r).castToI64().or_(Literal(int64_t(0x8000000000000LL))).castToF64();
       }

